optimize simple min aggregates.

aeflores opened the pull request:
Optimize aggregates such as the following:
```
.decl A(X:number,Y:number)
.decl next(X:number, Y:number)

next(X,Z):- A(X,_),  Z = min Y: { A(Y,_),Y>X}.
```
When we iterate over the indexed range in the aggregate, we know that the first element is already guaranteed to be the minimum, so we can leave the loop early. There is no need to traverse the whole range.

This makes the cost of the aggregate above basically constant.

A similar thing could be done with `max` if we could iterate backwards.

mmcgr commented:
The linter suggests a slight simplification:
```diff
diff --git a/src/synthesiser/Synthesiser.cpp b/src/synthesiser/Synthesiser.cpp
index f3eb108..3b7933f 100644
--- a/src/synthesiser/Synthesiser.cpp
+++ b/src/synthesiser/Synthesiser.cpp
@@ -1136,12 +1136,9 @@ void Synthesiser::emitCode(std::ostream& out, const Statement& stmt) {
         bool isGuaranteedToBeMinimum(const IndexAggregate& aggregate) {
             auto identifier = aggregate.getTupleId();
             auto keys = isa->getSearchSignature(&aggregate);
-            const TupleElement* tupleElem = dynamic_cast<const TupleElement*>(&aggregate.getExpression());
-            if (tupleElem && tupleElem->getTupleId() == identifier &&
-                    keys[tupleElem->getElement()] != ram::analysis::AttributeConstraint::None) {
-                return true;
-            }
-            return false;
+            const auto* tupleElem = dynamic_cast<const TupleElement*>(&aggregate.getExpression());
+            return tupleElem && tupleElem->getTupleId() == identifier &&
+                   keys[tupleElem->getElement()] != ram::analysis::AttributeConstraint::None;
         }
         void visitIndexAggregate(const IndexAggregate& aggregate, std::ostream& out) override {
             PRINT_BEGIN_COMMENT(out);
```

b-scholz commented:
The linter is complaining:
```
+ git diff -m
diff --git a/src/synthesiser/Synthesiser.cpp b/src/synthesiser/Synthesiser.cpp
index f3eb108..3b7933f 100644
--- a/src/synthesiser/Synthesiser.cpp
+++ b/src/synthesiser/Synthesiser.cpp
@@ -1136,12 +1136,9 @@ void Synthesiser::emitCode(std::ostream& out, const Statement& stmt) {
         bool isGuaranteedToBeMinimum(const IndexAggregate& aggregate) {
             auto identifier = aggregate.getTupleId();
             auto keys = isa->getSearchSignature(&aggregate);
-            const TupleElement* tupleElem = dynamic_cast<const TupleElement*>(&aggregate.getExpression());
-            if (tupleElem && tupleElem->getTupleId() == identifier &&
-                    keys[tupleElem->getElement()] != ram::analysis::AttributeConstraint::None) {
-                return true;
-            }
-            return false;
+            const auto* tupleElem = dynamic_cast<const TupleElement*>(&aggregate.getExpression());
+            return tupleElem && tupleElem->getTupleId() == identifier &&
+                   keys[tupleElem->getElement()] != ram::analysis::AttributeConstraint::None;
         }
         void visitIndexAggregate(const IndexAggregate& aggregate, std::ostream& out) override {
             PRINT_BEGIN_COMMENT(out);
+ git diff-index --quiet HEAD --
```

SamArch27 commented:
Just want to comment that this idea is excellent. We also need to add some extra checks.

We have a guarantee that this transformation is correct for min only in the case where the aggregate expression is over a single attribute.  The lex-order of the B-Tree relative to the search must be only that attribute, and thus the storage of tuples is ascending by that attribute value. The check is already present in the PR so far. 

However, we also need to add a check that a B-Tree index stores the relation to be completely safe. Suppose we are using a Brie or EqRel which do not natively support indexed inequalities. In that case, the tuples may be stored in any order and then filtered, failing to guarantee the first result will be the minimum.

We need to add a check that the type of the attribute iterated is "number" or that if the type is not "number" ("float" or "unsigned") that we are not running on interpreter mode. Since the interpreter does not support indexed inequalities for unsigned/float types, tuples are not guaranteed to be stored, ascending on the chosen attribute. The inequality semantics are guaranteed by applying filter operations on the inequality predicate one at a time.

Required Checks:
1. B-Tree index is used for the relation
2. The attribute type is "number" otherwise we cannot be running on interpreter mode.

Regarding the "max" idea, we would need to use reverse iterators since each attribute is always stored in ascending order relative to its position in the lex-order.

aeflores commented:
@SamArch27  one note regarding 
> The lex-order of the B-Tree relative to the search must be only that attribute

I think the lex-order can include other attributes indexed with equality (see `next_in_set` in the added example). The inequality constraint has to be always the last in the lex-order, so I think we are safe there.

Apart from that:
1. I added the check for the btree.
2. I think I need your help with the second check. These changes are in the `synthesizer`. How can this affect the interpreter? I am not familiar enough with the code. How can I check if I am in interpreter mode?

For `max`, would it be enough to implement `iterator& operator--()` for `Btree.h` and generate something like:
```
for(auto it=rel.end(); it!=rel.begin();it-- )
const auto& env = it*;
```
instead of `for(const auto& env  : rel)`?

SamArch27 commented:
> @SamArch27 one note regarding
> 
> > The lex-order of the B-Tree relative to the search must be only that attribute
> 
> I think the lex-order can include other attributes indexed with equality (see `next_in_set` in the added example). The inequality constraint has to be always the last in the lex-order, so I think we are safe there.
> 
> Apart from that:
> 
> 1. I added the check for the btree.
> 2. I think I need your help with the second check. These changes are in the `synthesizer`. How can this affect the interpreter? I am not familiar enough with the code. How can I check if I am in interpreter mode?
> 
> For `max`, would it be enough to implement `iterator& operator--()` for `Btree.h` and generate something like:
> 
> ```
> for(auto it=rel.end(); it!=rel.begin();it-- )
> const auto& env = it*;
> ```
> 
> instead of `for(const auto& env : rel)`?

Since we have an equality constraint on Set and an inequality constraint on Elem our SearchSignature should be 12. As you mentioned, attributes with an indexed inequality constraint come last in the lex-order relative to the search. For this example,
the B-Tree can only be ordered using the lex-order 0 < 1 (attribute 0 before attribute 1) or (Set < Elem).

The tuples will not be ordered by ascending Elem because Set has priority.

Consider for example the relation R = { <0, 10>, <10, 0> } we have <0, 10> less than <10, 0> in the lex-order because the Set attribute is compared first.

If we want to guarantee that the smallest value of an attribute appears first in a range, then it must be the first attribute in the lex-order relative to that search. When dealing with indexed inequalities if we want to query for the attribute which has the inequality constraint, this only occurs when this is attribute is the only one in the aggregate expression.

Regarding the interpreter check, since we are only adding this transformation for AggregateOp::MIN then we are safe even in interpreter mode because this aggregate only operates on number types. The interpreter only has issues with indexed inequalities for unsigned and float types (AggregateOp::UMIN and AggregateOp::FMIN).

I think adding reverse iterators would allow max to work. I am not familiar with the B-Tree implementation, and how easy it would be to add these. @b-scholz and @mmcgr are more familiar with this.

SamArch27 commented:
@aeflores: I realised why the next_in_set example works correctly.

    next_in_set(Set,Elem,Elem2):- set(Set,Elem), Elem2 = min E:{ set(Set,E), E > Elem}.

The reason why this works is that the type of attribute "Set" is "symbol". In Souffle, we don't apply indexed equalities on strings or "symbol" attributes. We naively scan and filter for the matching symbol.

Therefore, the SearchSignature is not 12 as I speculated in the previous comment. In fact, it is 02. 
This gives us the lex-order of 1 (compare the 1st attribute only). Therefore, we are in the same case as before as if the aggregate only has the attribute with an inequality constraint.

I think if we want to handle all of the cases explicitly you can just get the SearchSignature and check that the SearchSignature is all 0 with a single 2 (i.e. all values are AttributeConstraint::None with a single AttributeConstraint::Inequal). 

Hope that clears things up!

aeflores commented:
Hi @SamArch27 
> The tuples will not be ordered by ascending Elem because Set has priority.

This is true, but this is precisely what we want!!

In my example, the tuples will be ordered as follows in the index 12:
```
("even",2) <
("even",4) <
("even",6) <
("even",8) <
("even",10) <
("odd",1) <
("odd",3) <
("odd",5) <
("odd",7) <
("odd",9)
```
Obviously, the tuples are not sorted by ascending elem e.g. ("even",2) is smaller than ("odd",1). But the query is fixing the set, so the range query will return a range  where the Set is fixed e.g. Set = "odd" and within that fixed set the elems are in the right order:
```
("odd",1) <
("odd",3) <
("odd",5) <
("odd",7) <
("odd",9)
```
We are getting the next element for a fixed `Set`, and for that, the index `12` is precisely what we need.

SamArch27 commented:
> Hi @SamArch27
> 
> > The tuples will not be ordered by ascending Elem because Set has priority.
> 
> This is true, but this is precisely what we want!!
> 
> In my example, the tuples will be ordered as follows in the index 12:
> 
> ```
> ("even",2) <
> ("even",4) <
> ("even",6) <
> ("even",8) <
> ("even",10) <
> ("odd",1) <
> ("odd",3) <
> ("odd",5) <
> ("odd",7) <
> ("odd",9)
> ```
> 
> Obviously, the tuples are not sorted by ascending elem e.g. ("even",2) is smaller than ("odd",1). But the query is fixing the set, so the range query will return a range where the Set is fixed e.g. Set = "odd" and within that fixed set the elems are in the right order:
> 
> ```
> ("odd",1) <
> ("odd",3) <
> ("odd",5) <
> ("odd",7) <
> ("odd",9)
> ```
> 
> We are getting the next element for a fixed `Set`, and for that, the index `12` is precisely what we need.

Yep, I edited my comment just as you posted this. You're 100% correct here.

aeflores commented:
> Regarding the interpreter check, since we are only adding this transformation for AggregateOp::MIN then we are safe even in interpreter mode because this aggregate only operates on number types. The interpreter only has issues with indexed inequalities for unsigned and float types (AggregateOp::UMIN and AggregateOp::FMIN).

Actually, this optimization is being applied to `FMIN` and `UMIN` too (they are considered together in the same case). I would really like to have this for `UMIN` even if it is for only synthesized code. How does the interpreter use synthesized code?

aeflores commented:
I extended the test for Brie and `UMIN`. The transformation is not applied to Brie but it is applied to `UMIN`.

ohamel-softwaresecure commented on tests/semantic/aggregate4/aggregate4.dl:
```diff
@@ -0,0 +1,83 @@
+uset("odd",1).
+uset("odd",3).
+
+.decl next_in_uset(Set:symbol,Elem:unsigned,Elem2:unsigned)
+.output next_in_uset
+
+next_in_uset(Set,Elem,Elem2):- uset(Set,Elem),
+                               Elem2 = min E:{ uset(Set,E), E > Elem }.
```
Can't apply this to anything other than number b/c BTree is ordering on RamDomain (AKA RamSigned) interpretation of the bitpattern. It'll fail for every other primitive type, including unsigned.

e.g. 0xFFFFFFFFu will be ordered before 0u since that's -1 when interpreted as 2's complement RamSigned.

Other types:
Symbols: completely unordered.
Floats: *laughs in equiv bit-patterns*

aeflores replied:
Are you sure about this? I added `0xFFFFFFFF` to the unsigned example and it seems to work well.

Also, I thought the comparators for the btrees cast to the corresponding domain before comparing.

SamArch27 replied:
@b-scholz and I have worked together on the type comparators for B-Trees so float attributes are compared semantically rather than by bit-value. @ohamel-softwaresecure have you found a test case which fails?

ohamel-softwaresecure replied:
Oops. Alright ignore me. I'm trapped between a divergent fork of souffle and upstream. The old fork didn't do that before.

SamArch27 commented:
> > Regarding the interpreter check, since we are only adding this transformation for AggregateOp::MIN then we are safe even in interpreter mode because this aggregate only operates on number types. The interpreter only has issues with indexed inequalities for unsigned and float types (AggregateOp::UMIN and AggregateOp::FMIN).
> 
> Actually, this optimization is being applied to `FMIN` and `UMIN` too (they are considered together in the same case). I would really like to have this for `UMIN` even if it is for only synthesized code. How does the interpreter use synthesized code?

I've double-checked with @XiaowenHu96 and since your transformation is in the synthesiser, it won't affect the interpreter. In this case, you can assume that if a B-Tree index is leveraged then attributes w/inequalities (float or unsigned) will also be indexed.

aeflores commented:
@SamArch27 I think I addressed all your comments for the `min` implementation. Is there anything else?
Is there any other test that we should add?

 I would rather leave the corresponding `max` optimization for a different merge request.

SamArch27 commented:
> @SamArch27 I think I addressed all your comments for the `min` implementation. Is there anything else?
> Is there any other test that we should add?
> 
> I would rather leave the corresponding `max` optimization for a different merge request.

I have discussed with @b-scholz and we can actually speed up a larger class of min/max aggregates. Consider the aggregate:
    new(z) :- other(y), z = min x : { rel(x,y) }

The SearchSignature of the aggregate is 01, so our evaluation of the min aggregate currently will be a naive table scan over the range for the min. However, we can use a clever trick where at the index selection stage we force the SearchSignature to be 21. In effect, this will make the attribute x in the resultant range ascending. We then call lowerUpperRange_21(...), and we can guarantee that the first tuple in the range minimises x.

This technique is more general, but I'm quite busy finishing my honours thesis so I don't think I can implement this until December when I have more time. 

In my view, I think this PR has all of the necessary tests, so it is safe to merge as-is. I think in a future PR we can always add the technique I described above.

@b-scholz will have to confirm if we are safe to merge the PR now or whether we should wait and implement the more general min strategy right now.

b-scholz commented:
Sorry for joining the discussion late. Here are my observations:

(1) A more general approach would be to identify aggregates of the format 
        `y = min/max x: { A(z1, ..., zk-1, x, zk+1, ... zn)}` 
     where  the arguments  `z1 ... zn` are arbitrary arguments (variables, constants, expressions, etc.) and add a search signature in the IndexAnalyssi for such aggregates. The advantage would be that aggregates of this nature would always obtain an index. 

(2) The current implementation does not use directly the lower-bound of the range; it breaks out of the loop after the first iteration. As a consequence, a lower and upper bound is computed. It would be better to use the lower bound for min only. Similarly, for max-aggregates, we would like to employ a specific upper bound operation. 

Perhaps we merge the current approach to master and mark the implementation as work in progress.
